sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -350,7 +384,7 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
       literalType: DataType): Boolean = {
     toType.sameType(literalType) &&
       !fromExp.foldable &&
-      toType.isInstanceOf[NumericType] &&
+      (toType.isInstanceOf[NumericType] || toType.isInstanceOf[DateType]) &&
```

Comment:
do we need to change here? The new code does not call `canImplicitlyCast`

Reply:
+1

Reply:
@cloud-fan @sunchao This is used to support `literal BinaryComparator cast(exp as timesatmp)`, e.g.: `DATE '2023-01-01' < CAST(ts AS date)`. I have moved the logic to a new function: `needToCanonicalizeExpression`.